File: src/main/scala/com/vc/medium/L227.scala
package com.vc.medium

object L227 {
  def calculate(s: String): Int = {
    val n = s.length
    var i = 0
    import scala.collection.mutable
    val st = new mutable.Stack[Int]()
    var num = 0
    while(i < n) {
      val ch = s(i)
      if(Character.isDigit(ch)) {
        num = num * 10 + (ch - '0')
      }
      else if(ch == '+') {
        st.push(num)
        num = 0
      }
      else if(ch == '-') {
        st.push(-1 * num)
        num = 0
      }
      else if(ch == '*') {
        st.push(st.pop * num)
        num = 0
      }
      else if(ch == '/') {
        st.push(st.pop / num)
      }
      i += 1
    }
    var res = 0
    while(st.nonEmpty) {
      res += st.pop
    }
    res
  }

  def main(args: Array[String]): Unit = {
    calculate("3+2*2")
  }
}

File: src/main/scala/com/vc/medium/L227.java
package com.vc.medium;

import java.util.*;

class L227 {
    public int calculate(String s) {
        int sum = 0;
        int i = 0;

        Set<Character> set = new HashSet<Character>();
        set.add('+');
        set.add('-');
        set.add('*');
        set.add('/');

        Stack<Integer> st = new Stack<Integer>();
        char prevSign = '+';
        int n = s.length();

        int num = 0;
        while(i < n) {
            char ch = s.charAt(i);
            if(Character.isDigit(ch)) {
                num = num * 10 + (ch - '0');
            }
            if(set.contains(ch) || i == n - 1) {
                if(prevSign == '+') {
                    st.push(num);
                }
                else if(prevSign == '-') {
                    st.push(-1 * num);
                }
                else if(prevSign == '*') {
                    st.push(st.pop() * num);
                }
                else if(prevSign == '/') {
                    st.push(st.pop() / num);
                }
                prevSign = ch;
                num = 0;
            }
            i++;
        }
        while(!st.isEmpty()) {
            sum += st.pop();
        }
        return sum;
    }
}